### PR TITLE
Isaicp 3682: Limit headings in the content editor's UI

### PR DESCRIPTION
**Pull request by idimopoulos**

### CODE REVIEW

File: web/profiles/joinup/joinup.behat.inc [hunk 1]
```diff
@@ -2006,4 +2006,55 @@ class JoinupSubContext extends DrupalSubContextBase implements DrupalSubContextI
...
+      var element_id = arguments[0]['element_id'];
+      return CKEDITOR.instances[element_id].config.format_tags;
+JS;
+    $found_tags = $this->getSession()->getDriver()->getWebDriverSession()->execute([
+      'script' => $has_tags_condition,
+      'args' => [['element_id' => $element_id]],
+    ]);
+    $found_tags_array = explode(';', $found_tags);
+    if (array_intersect($format_tags, $found_tags_array)) {
+      throw new Exception("Some tags were found in the ckeditor.");
```
Comment (brummbar) [inline web/profiles/joinup/joinup.behat.inc]: `\Exception`. Also print the one founds. And double quotes.

File: web/profiles/joinup/joinup.behat.inc [hunk 2]
```diff
@@ -2006,4 +2006,55 @@ class JoinupSubContext extends DrupalSubContextBase implements DrupalSubContextI
...
+        }
+      }
+    }
+    throw new \Exception(sprintf('The text "%s" was not found in the "%s" element in the "%s" region on the page %s', $text, $tag, $region, $this->getSession()->getCurrentUrl()));
+  }
+
+  /**
+   * Asserts that a ckeditor list contains an element.
+   *
+   * This step requires javascript.
```
Comment (brummbar) [inline web/profiles/joinup/joinup.behat.inc]: Then use `\Drupal\joinup\Traits\BrowserCapabilityDetectionTrait::browserSupportsJavascript()`.

File: web/profiles/joinup/joinup.behat.inc [hunk 3]
```diff
@@ -2006,4 +2006,55 @@ class JoinupSubContext extends DrupalSubContextBase implements DrupalSubContextI
...
+   * CKeditor stores the lists available in the header of the editor pane in an
+   * iframe with a role tag. The dropdown "Format" has to be clicked prior to
+   * having the iframe available.
+   *
+   * @Then the paragraph formats in the :field field should not contain the :format_tags format
+   */
+  public function assertNotFormatInCkeditorExists($field, $format_tags) {
+    $element = $this->getSession()->getPage()->findField($field);
+    $element_id = $element->getAttribute('id');
+    $format_tags = array_map('trim', explode(',', $format_tags));
```
Comment (brummbar) [inline web/profiles/joinup/joinup.behat.inc]: Please use `\Drupal\joinup\Traits\UtilityTrait::explodeCommaSeparatedStepArgument()`.

File: web/profiles/joinup/joinup.behat.inc [hunk 4]
```diff
@@ -2006,4 +2006,55 @@ class JoinupSubContext extends DrupalSubContextBase implements DrupalSubContextI
...
+    $has_tags_condition = <<<JS
+      var element_id = arguments[0]['element_id'];
+      return CKEDITOR.instances[element_id].config.format_tags;
```
Comment (brummbar) [inline web/profiles/joinup/joinup.behat.inc]: There is no need to pass an argument here. You could change the condition to:
`return CKEDITOR.instances[$element_id].config.format_tags;`
making the script passed to the browser already prepared.

File: web/profiles/joinup/joinup.behat.inc [hunk 5]
```diff
@@ -2006,4 +2006,55 @@ class JoinupSubContext extends DrupalSubContextBase implements DrupalSubContextI
...
   }
 
+  /**
+   * Asserts that an tag with the given text exists in the given region.
+   *
+   * This completes the assertNotRegionElementText of Markup context.
+   *
+   * @see \Drupal\DrupalExtension\Context\MarkupContext::assertNotRegionElementText()
+   *
+   * @Then I( should) see :text in the :tag element in the :region( region)
```
Comment (brummbar) [inline web/profiles/joinup/joinup.behat.inc]: I think the step could be improved a bit, as actually it checks that any :tag element has the wanted text, and there might be more than one :tag in the region.
`@Then I( should) see a :tag (HTML )element with the text :text in the :region( region)`

File: web/profiles/joinup/joinup.behat.inc [hunk 6]
```diff
@@ -2006,4 +2006,55 @@ class JoinupSubContext extends DrupalSubContextBase implements DrupalSubContextI
...
+   */
+  public function assertRegionElementText($text, $tag, $region) {
+    $regionObj = $this->getRegion($region);
+    $results = $regionObj->findAll('css', $tag);
+    if (!empty($results)) {
+      foreach ($results as $result) {
+        if ($result->getText() == $text) {
```
Comment (brummbar) [inline web/profiles/joinup/joinup.behat.inc]: Use `===`.

File: web/modules/custom/joinup_core/joinup_core.module [hunk 7]
```diff
@@ -929,3 +929,17 @@ function joinup_core_form_views_exposed_form_alter(&$form, FormStateInterface $f
 function joinup_core_preprocess_page(&$variables) {
   $variables['#attached']['library'][] = 'joinup_core/sign_in_destination';
 }
+
+/**
+ * Implements hook_editor_js_settings_alter().
+ *
+ * Strip off the h1, h5, h6 paragraph format options from the 'Format' dropdown.
```
Comment (brummbar) [inline web/modules/custom/joinup_core/joinup_core.module]: Swap "paragraph" here with tags, as they actually are tags.

File: web/modules/custom/joinup_core/joinup_core.module [hunk 8]
```diff
@@ -929,3 +929,17 @@ function joinup_core_form_views_exposed_form_alter(&$form, FormStateInterface $f
...
+ * This does not remove the ability of the user to enter or render these tags.
+ *
+ * @see https://webgate.ec.europa.eu/CITnet/jira/browse/ISAICP-3682
+ */
+function joinup_core_editor_js_settings_alter(array &$settings) {
+  if (isset($settings['editor']['formats']['content_editor']) && $settings['editor']['formats']['content_editor']['format'] === 'content_editor') {
+    $settings['editor']['formats']['content_editor']['editorSettings']['format_tags'] = preg_replace('/h[1|5|6].*?;/', '', $settings['editor']['formats']['content_editor']['editorSettings']['format_tags']);
```
Comment (brummbar) [inline web/modules/custom/joinup_core/joinup_core.module]: We discussed slightly this already, let's improve this regex, maybe we can explode items to be sure that we don't catch tags that end in `h[1|5|6]`.

File: web/modules/custom/joinup_core/joinup_core.module (hunk 7 above)
Comment (idimopoulos) [reply web/modules/custom/joinup_core/joinup_core.module]: I changed it with 'tag' as the plural goes to options.

File: web/profiles/joinup/joinup.behat.inc [hunk 9]
```diff
@@ -2006,4 +2006,55 @@ class JoinupSubContext extends DrupalSubContextBase implements DrupalSubContextI
...
+  /**
+   * Asserts that an tag with the given text exists in the given region.
+   *
+   * This completes the assertNotRegionElementText of Markup context.
+   *
+   * @see \Drupal\DrupalExtension\Context\MarkupContext::assertNotRegionElementText()
+   *
+   * @Then I( should) see a(n) :tag element with the text :text in the :region( region)
+   */
+  public function assertRegionElementText($tag, $text, $region) {
```
Comment (pfrenssen) [inline web/profiles/joinup/joinup.behat.inc]: Parameters are not documented.

File: web/profiles/joinup/joinup.behat.inc [hunk 10]
```diff
@@ -2006,4 +2006,55 @@ class JoinupSubContext extends DrupalSubContextBase implements DrupalSubContextI
...
+  /**
+   * Asserts that a ckeditor list contains an element.
+   *
+   * CKeditor stores the lists available in the header of the editor pane in an
+   * iframe with a role tag. The dropdown "Format" has to be clicked prior to
+   * having the iframe available.
+   *
+   * @Then the paragraph formats in the :field field should not contain the :format_tags format
+   */
+  public function assertNotFormatInCkeditorExists($field, $format_tags) {
```
Comment (pfrenssen) [inline web/profiles/joinup/joinup.behat.inc]: Parameters are not documented, and the `$format_tags` accept multiple comma separated values, so I would change the step definition to `... format(s)`.

**Comment by pfrenssen** [comment]
Tests were not started automatically, I started them manually: https://app.continuousphp.com/git-hub/ec-europa/joinup-dev/build/f78f790a-038f-4af5-af69-a0f0853a9904

**Pull request merged**